src/Microsoft.AspNet.OData.Shared/DeltaSetOfT.cs

```diff
@@ -0,0 +1,357 @@
+﻿// Copyright (c) Microsoft Corporation.  All rights reserved.
+// Licensed under the MIT License.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Diagnostics.Contracts;
+using System.Linq;
+using System.Reflection;
+using Microsoft.AspNet.OData.Builder;
+using Microsoft.AspNet.OData.Common;
+using Org.OData.Core.V1;
+
+namespace Microsoft.AspNet.OData
+{
+    /// <summary>
+    /// Represents an <see cref="IDeltaSet"/> that is a collection of <see cref="IDeltaSetItem"/>s.
+    /// </summary>
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Naming", "CA1710:IdentifiersShouldHaveCorrectSuffix")]
+    [NonValidatingParameterBinding]
+    public class DeltaSet<TStructuralType> : Collection<IDeltaSetItem>, IDeltaSet where TStructuralType : class
+    {        
+        private Type _clrType;
+        IList<string> _keys;
+               
+        internal IDictionary<string, IDictionary<string, object>> dictKeyValues;
+        internal string currentPathValue;
+
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="DeltaSet{TStructuralType}"/> class.
+        /// </summary>
+        public DeltaSet()            
+        {            
+            _clrType = typeof(TStructuralType);
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="DeltaSet{TStructuralType}"/> class.
+        /// </summary>
+        /// <param name="keys">List of key names for the type</param>
+        public DeltaSet(IList<string> keys)           
+        {
+            _keys = keys;            
+            _clrType = typeof(TStructuralType);            
+        }
+
+    
+       /// <inheritdoc/>      
+        protected override void InsertItem(int index, IDeltaSetItem item)
+        {
+            Delta<TStructuralType> deltaItem = item as Delta<TStructuralType>;
+
+            if (deltaItem == null)
+            {
+                throw Error.Argument("item", SRResources.ChangedObjectTypeMismatch, item.GetType(), typeof(TStructuralType));
+            }
+
+            base.InsertItem(index, item);
+        }
+
+        /// <summary>
+        /// Handler for users Create, Get and Delete Methods
+        /// </summary>
+        internal PatchMethodHandler<TStructuralType> PatchHandler { get; set; }
```

Comment:
>PatchHandler [](start = 53, length = 12)

see related comments. I would rather pass PatchHandler ot the Patch/CopyChangedValues method than have it an internal property on the DeltaSet<T>